fix/longitudinal_distance_fixed

lchojnack opened the pull request:
## Types of PR

- [ ] New Features
- [ ] Upgrade of existing features
- [x] Bugfix

## Link to the issue

## Description
The PR proposes four minor changes for https://github.com/tier4/scenario_simulator_v2/pull/924 in order to pass CI regression tests
Noticed issues and how they are resolved:
1. Wrong calculation of longitudinal distance -> da0abb2d190f39769ba58d3401c3e5b910bd231e
In [`EntityManager::getLongitudinalDistance()` ](https://github.com/tier4/scenario_simulator_v2/blob/5fff6171034dd2112e74430001ce32972a412913/simulation/traffic_simulator/src/entity/entity_manager.cpp#L252) function on `master` branch the calculation of longitudinal distance is the following
```c++
  if (forward_distance && backward_distance) {
    if (forward_distance.value() > backward_distance.value()) {
      return -backward_distance.value();
    } else {
      return forward_distance.value();
    }
  }
```
and in https://github.com/tier4/scenario_simulator_v2/pull/924 it was changed to
```c++
if (forward_distance && backward_distance) {
      return std::min(forward_distance.value(), std::abs(backward_distance.value()));
 }
```
The change does not take into account the `-` sign. The commit da0abb2d190f39769ba58d3401c3e5b910bd231e reverts the calculation to the version from `master` branch.

2. Finishing `WalkStraightAction` by Pedestrian NPC -> 8d5adb93543bd6a2c065730c4011437ec86f368b
In the [PedestrianEntity::onUpdate()](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L227) function the stopping of pedestrian `stopAtCurrentPosition()` is done after [setting the status and calling update](https://github.com/tier4/scenario_simulator_v2/blob/fix/longitudinal_distance/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L231-L237). When Pedestrian NPC is close to the end of lanelet (e.g s=9.99), lanelet length is s=10m, and [behavior_plugin_ptr_->update()](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L237) is calling, then function [CanonicalizedLaneletPose::canonicalize()](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/data_type/lanelet_pose.cpp#L39) is used on updated position (e.g s=10.01). For that reason some tests fails with the message `Simulation error [SemanticError]: /home/autoware/autoware.proj/src/simulator/scenario_simulator/simulation/traffic_simulator/src/data_type/lanelet_pose.cpp:48: Lanelet pose (id=1595,s=9.97524,offset=0.212264,rpy.x=0,rpy.y=0,rpy.z=-3.12087) is invalid, please check lanelet length and connection.` or similar.  Commit 8d5adb93543bd6a2c065730c4011437ec86f368b adds a small offset in order to allow a position (e.g. s=10.01) when lanelet length (e.g. s=10.0) without throwing the exception.

3. Wrong `getFollowingLanelets()` lanelets return -> adb9c0f4f27a8db09843849eb4eaaeaa4eea68f7
In the [HdMapUtils::getFollowingLanelets()](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L715) function the following lanelets are return as follows
```c++
  return getFollowingLanelets(
    candidate_lanelet_ids[candidate_lanelet_ids.size() - 1], 
    distance - total_distance, false) + ret;
```
and on `master` branch [a similar code is](https://github.com/tier4/scenario_simulator_v2/blob/master/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L661-L664)
```c++
  std::int64_t end_lanelet = candidate_lanelet_ids[candidate_lanelet_ids.size() - 1];
  const auto followings = getFollowingLanelets(end_lanelet, distance - total_distance, false);
  std::copy(followings.begin(), followings.end(), std::back_inserter(ret));
  return ret;
```
The `std::back_inserter()` is [used to add elements to the end of the container](https://en.cppreference.com/w/cpp/iterator/back_inserter), so the `followings` container is added to `ret` container. 
In `fix/longitudinal_distance` branch the `ret` container is added to the `following` container which causes an issue in getting a route for Vehicle NPC. Commit adb9c0f4f27a8db09843849eb4eaaeaa4eea68f7 provides a fix and adds  `followings` container to `ret` container.

4. Canonicalizing LaneletPose when `s < 0` -> 71ba3935f64df600e7508370cd40ef81a31133e1
The cause of regression has been noticed in the negative report below
o⇒x : UC-A-01_通常走行_経路追従#2_zero / UC-A-01-00005-001-case2z_v2
base : [URL](https://evaluation.tier4.jp/evaluation/reports/e314aeb6-4c48-50b2-97c4-d0692873da19/tests/9f17ad34-11da-5cfb-9975-df4c98824276/2f9b7e2c-9dba-579c-a0a2-619bcda6c02e/12addf49-03c8-5544-b447-d4bf0334a9d1?project_id=odd2_reference)
eval : [URL](https://evaluation.tier4.jp/evaluation/reports/4933c3c4-ff50-523c-ba31-436bd0b18ab8/tests/8ca78eed-9434-5f83-8cd9-8c3d94323d15/910397b4-e22f-5df7-9024-496cd15b64c2/28adbf28-79b2-5506-af39-ffedd1b1ad92?project_id=odd2_reference)
Params : {}
Message : Simulation failure: Expected success

In this scenario there is `Point3` located in (lane=1297,s=-1.865)
<img width="958" alt="image" src="https://github.com/tier4/scenario_simulator_v2/assets/8416175/10f30552-2617-48f8-bb1b-ff1803920418">

The `Point3` has two previous lanelets (lane=1375 and lane=1290). The lanelet pose is canonicalized only to the first lanelet ([code](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#LL98C1-L101C6)). For that reason `forward_distance`  and `backward_distance` in function `EntityManager::getLongitudinalDistance` could be wrongly calculated ([code](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/entity/entity_manager.cpp#LL256C1-L260C69)).

When `Point3` is canonicalized to the lane=1375 then the distance between `EGO` and `Point3` is computed as marked in the yellow line in the picture below. If  `Point3` would be canonicalized to the lane=1290 then the distance between `EGO` and `Point3` would be computed as marked in the orange line.
![image](https://github.com/tier4/scenario_simulator_v2/assets/8416175/8fe0f3e2-b8cb-4bbb-91d9-f47ae085b38c)

In the `master` branch the longitudinal distance for this particular case is computed as marked in the orange line, but in the `fix/longitudinal_distance` is computed as marked in the yellow line. 

For that reason, the distance condition is `true` for `eval` and that leads to the failure of the scenario (presented in the screenshots below).  

| Baseline | Eval |
| ------------ | ------ |
| ![image](https://github.com/tier4/scenario_simulator_v2/assets/8416175/dea390ea-3131-4578-9e81-41789487b074) | ![image](https://github.com/tier4/scenario_simulator_v2/assets/8416175/d2a45f3c-179f-48e9-bded-a81a9df06601) |

The solution could be keeping the `non canonicalized lanelet pose` in the `CanonicalizedLaneletPose` class and using it for the computation of `forward_distance`  and `backward_distance`. The commit 71ba3935f64df600e7508370cd40ef81a31133e1 implements this solution


## How to review this PR.

hakuturu583 commented on simulation/traffic_simulator/include/traffic_simulator/data_type/lanelet_pose.hpp:
```diff
@@ -66,6 +67,7 @@ Offset and rpy values are not taken into account.
     const std::shared_ptr<hdmap_utils::HdMapUtils> & hdmap_utils,
     const std::vector<std::int64_t> & route_lanelets) -> LaneletPose;
   const LaneletPose lanelet_pose_;
+  LaneletPose non_canonicalized_lanelet_pose;
```
Is this member variable is necessary?

hakuturu583 commented on simulation/traffic_simulator/include/traffic_simulator/data_type/lanelet_pose.hpp:
```diff
@@ -35,6 +35,7 @@ class CanonicalizedLaneletPose
     const std::vector<std::int64_t> & route_lanelets);
   explicit operator LaneletPose() const noexcept { return lanelet_pose_; }
   explicit operator geometry_msgs::msg::Pose() const noexcept { return map_pose_; }
+  LaneletPose getNonCanonicalizedLaneletPose() const { return non_canonicalized_lanelet_pose; }
```
Is this member function is necessary?

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_manager.cpp:
```diff
@@ -254,13 +254,17 @@ auto EntityManager::getLongitudinalDistance(
 
     auto backward_distance = hdmap_utils_ptr_->getLongitudinalDistance(
-      static_cast<LaneletPose>(to), static_cast<LaneletPose>(from));
+      to.getNonCanonicalizedLaneletPose(), from.getNonCanonicalizedLaneletPose());
```
We believe that after the LaneletPose is Canonicalized, the Lanelet Pose before Canonicalization should no longer be needed.
The fix included in the /fix/longitudinal_distance branch fixes a bug that was caused by the Lanelet Pose before it was Canonicalized and allows the NonCanonicalizedLaneletPose to be used with the If you use NonCanonicalizedLaneletPose to calculate longitudinal distance, the bug will not be corrected.

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_manager.cpp:
```diff
@@ -254,13 +254,17 @@ auto EntityManager::getLongitudinalDistance(
   if (!include_adjacent_lanelet) {
     auto forward_distance = hdmap_utils_ptr_->getLongitudinalDistance(
-      static_cast<LaneletPose>(from), static_cast<LaneletPose>(to));
+      from.getNonCanonicalizedLaneletPose(), to.getNonCanonicalizedLaneletPose());
```
We believe that after the LaneletPose is Canonicalized, the Lanelet Pose before Canonicalization should no longer be needed.
The fix included in the /fix/longitudinal_distance branch fixes a bug that was caused by the Lanelet Pose before it was Canonicalized and allows the NonCanonicalizedLaneletPose to be used with the If you use NonCanonicalizedLaneletPose to calculate longitudinal distance, the bug will not be corrected.

lchojnack replied:
The reason, why I've decided to store the Lanelet Pose before Canonicalization and use it in the `hdmap_utils_ptr_->getLongitudinalDistance()` is that when the LaneletPose is Canonicalizated then [only first lanelet_id with the lowest array index is used as a candidate](https://github.com/tier4/scenario_simulator_v2/blob/fd4ff7a26613402134bf9737d6ef98e11d664b4b/simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp#L99). 
For a particular example, which is described at the 4. point in Description section in this PR, Canonicalization only to the first lanelet_id causes the scenario fails. 

Please, give me some time, I will try to find another solution for this issue.

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -100,7 +144,7 @@ auto HdMapUtils::canonicalizeLaneletPose(
     }
   }
-  while (canonicalized.s > getLaneletLength(canonicalized.lanelet_id)) {
+  while (canonicalized.s > (getLaneletLength(canonicalized.lanelet_id) + 0.2)) {
```
/// @note Hard coded parameter! (Please describe why the parameter is 0.2) and leave a comment explaining what this parameter does and why it should be that value.

lchojnack replied:
I've removed this hardcoded value because it didn't work for all cases.
I've added checking if a pose can be canonicalized in `PedestrianActionNode::estimateLaneletPose()` function
This change is in https://github.com/tier4/scenario_simulator_v2/pull/992/commits/1bba583ce23693331a4105b77480063bd325ec11

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
+          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), 0);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.push_back(lanelet_pose_tmp);
```
```suggestion
          canonicalized_all.emplace_back(lanelet_pose_tmp);
```

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
+          id, lanelet_pose.s + getLaneletLength(id), 0);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.push_back(lanelet_pose_tmp);
```
```suggestion
          canonicalized_all.emplace_back(lanelet_pose_tmp);
```

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s + getLaneletLength(id), 0);
```
It seems, this line overwrites offset values and push_back in line 102.
I think offset values should be copy from lanelet_pose.

```suggestion
          id, lanelet_pose.s + getLaneletLength(id), lanelet_pose.offset);
```

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), 0);
```
It seems, this line overwrites offset values and push_back in line 120.
I think offset values should be copy from lanelet_pose.

```suggestion
          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), lanelet_pose.offset);
```

hakuturu583 commented on simulation/traffic_simulator/test/src/test_hdmap_utils.cpp:
```diff
@@ -122,6 +122,116 @@ TEST(HdMapUtils, RoadShoulder)
   }
 }
 
+TEST(HdMapUtils, CanonicalizeNegative)
```
Please describe the test target and expected result in comment.

lchojnack replied:
Comment added

hakuturu583 commented on simulation/traffic_simulator/test/src/test_hdmap_utils.cpp:
```diff
@@ -122,6 +122,116 @@ TEST(HdMapUtils, RoadShoulder)
+}
+
+TEST(HdMapUtils, CanonicalizePositive)
+{
```
Please describe the test target and expected result in comment.

lchojnack replied:
Comment added

hakuturu583 commented on simulation/traffic_simulator/test/src/test_hdmap_utils.cpp:
```diff
@@ -122,6 +122,116 @@ TEST(HdMapUtils, RoadShoulder)
+}
+
+TEST(HdMapUtils, CanonicalizeAllPositive)
+{
```
Please describe the test target and expected result in comment.

lchojnack replied:
Comment added

lchojnack replied:
I've changed the comment format -> 177a344d7c55afa1d27d8946380ff00a468b40bd

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_manager.cpp:
```diff
@@ -253,14 +253,24 @@ auto EntityManager::getLongitudinalDistance(
 
     if (forward_distance && backward_distance) {
-      return std::min(forward_distance.value(), std::abs(backward_distance.value()));
+      if (forward_distance.value() > backward_distance.value()) {
+        return -backward_distance.value();
+      } else {
+        return forward_distance.value();
+      }
```
```suggestion
      return forward_distance.value() > backward_distance.value() ? -backward_distance.value() : forward_distance.value();
```

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_manager.cpp:
```diff
@@ -253,14 +253,24 @@ auto EntityManager::getLongitudinalDistance(
+    auto backward_distance =
+      hdmap_utils_ptr_->getLongitudinalDistance(to_canonicalized, static_cast<LaneletPose>(from));
 
     if (forward_distance && backward_distance) {
```
```suggestion
    if (const auto forward_distance = hdmap_utils_ptr_->getLongitudinalDistance(static_cast<LaneletPose>(from), to_canonicalized) && const auto backward_distance = hdmap_utils_ptr_->getLongitudinalDistance(to_canonicalized, static_cast<LaneletPose>(from))) {
```

lchojnack replied:
I'm not able to apply this code change suggestion because a few lines lower the `forward_distance` variable and the `backward_distance` variable is used in `else if ()` statement.

hakuturu583 replied:
I see. Thanks.

hakuturu583 commented on simulation/traffic_simulator/src/entity/entity_manager.cpp:
```diff
@@ -253,14 +253,24 @@ auto EntityManager::getLongitudinalDistance(
   if (!include_adjacent_lanelet) {
-    auto forward_distance = hdmap_utils_ptr_->getLongitudinalDistance(
-      static_cast<LaneletPose>(from), static_cast<LaneletPose>(to));
+    auto to_canonicalized = static_cast<LaneletPose>(to);
```
```suggestion
    const auto to_canonicalized = static_cast<LaneletPose>(to);
```

lchojnack replied:
I'm not able to add `const` to the `to_canonicalized` variable because if `to.hasAlternativeLaneletPose()` then assign to `to_canonicalized` alternative lanelet pose

hakuturu583 commented on simulation/traffic_simulator/src/data_type/lanelet_pose.cpp:
```diff
@@ -76,6 +80,23 @@ auto CanonicalizedLaneletPose::canonicalize(
+  -> LaneletPose
+{
+  std::vector<std::int64_t> shortest_route =
+    hdmap_utils->getRoute(from.lanelet_id, lanelet_poses_[0].lanelet_id);
```
Please care when the `lanelet_poses_.empty() == true`

lchojnack replied:
I've added 
```
 if (lanelet_poses_.empty()) {
    return std::nullopt;
  }
  ```

hakuturu583 commented:
@lchojnack Thanks, I review your code.

lchojnack commented:
@hakuturu583 Thank you for review. I've applied your feedback.
For https://github.com/tier4/scenario_simulator_v2/pull/992#discussion_r1234663346 I've removed the hardcoded value and I've added checking if a pose can be canonicalized in `PedestrianActionNode::estimateLaneletPose()`

hakuturu583 commented:
@lchojnack Please pass build test before requesting code review.

lchojnack commented:
> @lchojnack Please pass build test before requesting code review.

@hakuturu583 The build does not pass because of the error that is not related to the ss2 code but is related to the external library used in build test
<img width="690" alt="image" src="https://github.com/tier4/scenario_simulator_v2/assets/8416175/9beaee57-3f70-4b14-9acb-7285bb8cb151">


In PR https://github.com/tier4/scenario_simulator_v2/pull/1018 which is a PR https://github.com/tier4/scenario_simulator_v2/pull/992 with the merged `master` branch, all build test passes successfully. 

I think that fix for the issue with build test has been merged to master in PR https://github.com/tier4/scenario_simulator_v2/pull/1005

hakuturu583 commented:
Diff of the master branch is too big, so if you merge master first, I cannot review this branch.
So, please cherry-pick and fix build errors in this branch first.

hakuturu583 commented:
In my memory, remove BuildTest and use [BuildAndRun](https://github.com/tier4/scenario_simulator_v2/blob/master/.github/workflows/BuildAndRun.yaml) workflow in master branch can fix this problem.

hakuturu583 commented on simulation/traffic_simulator/src/hdmap_utils/hdmap_utils.cpp:
```diff
@@ -85,6 +85,50 @@ HdMapUtils::HdMapUtils(
+  -> std::vector<traffic_simulator_msgs::msg::LaneletPose>
+{
+  std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
+  if (lanelet_pose.s < 0) {
+    if (const auto ids = getPreviousLaneletIds(lanelet_pose.lanelet_id); ids.empty()) {
+      return canonicalized_all;
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s + getLaneletLength(id), lanelet_pose.offset);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.emplace_back(lanelet_pose_tmp);
+        } else {
+          std::copy(
+            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
+            std::back_inserter(canonicalized_all));
+        }
+      }
+    }
+  }
+  if (lanelet_pose.s > (getLaneletLength(lanelet_pose.lanelet_id))) {
+    if (const auto ids = getNextLaneletIds(lanelet_pose.lanelet_id); ids.empty()) {
+      return canonicalized_all;
+    } else {
+      for (const auto id : ids) {
+        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
+          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), lanelet_pose.offset);
+        if (auto canonicalized_lanelet_poses = gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
+            canonicalized_lanelet_poses.empty()) {
+          canonicalized_all.emplace_back(lanelet_pose_tmp);
+        } else {
+          std::copy(
+            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
+            std::back_inserter(canonicalized_all));
+        }
+      }
+    }
+  }
+  return canonicalized_all;
```
We believe that this writing style makes the structure of the function very difficult to grasp, since the If in lines 93 and 111 appear to be something that cannot be established at the same time.
We suggest that you rewrite the function as you see fit.

hakuturu583 replied:
```suggestion
  /// @note Define lambda functions for canonicalizing previous/next lanelet.
  const auto canonicalize_to_previous_lanelet =
    [this](const auto & lanelet_pose) -> std::vector<traffic_simulator_msgs::msg::LaneletPose> {
    if (const auto ids = getPreviousLaneletIds(lanelet_pose.lanelet_id); !ids.empty()) {
      std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
      for (const auto id : ids) {
        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
          id, lanelet_pose.s + getLaneletLength(id), lanelet_pose.offset);
        if (const auto canonicalized_lanelet_poses =
              gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
            canonicalized_lanelet_poses.empty()) {
          canonicalized_all.emplace_back(lanelet_pose_tmp);
        } else {
          std::copy(
            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
            std::back_inserter(canonicalized_all));
        }
      }
      return canonicalized_all;
    } else {
      return {};
    }
  };
  const auto canonicalize_to_next_lanelet =
    [this](const auto & lanelet_pose) -> std::vector<traffic_simulator_msgs::msg::LaneletPose> {
    if (const auto ids = getNextLaneletIds(lanelet_pose.lanelet_id); !ids.empty()) {
      std::vector<traffic_simulator_msgs::msg::LaneletPose> canonicalized_all;
      for (const auto id : ids) {
        const auto lanelet_pose_tmp = traffic_simulator::helper::constructLaneletPose(
          id, lanelet_pose.s - getLaneletLength(lanelet_pose.lanelet_id), lanelet_pose.offset);
        if (const auto canonicalized_lanelet_poses =
              gelAllCanonicalizedLaneletPoses(lanelet_pose_tmp);
            canonicalized_lanelet_poses.empty()) {
          canonicalized_all.emplace_back(lanelet_pose_tmp);
        } else {
          std::copy(
            canonicalized_lanelet_poses.begin(), canonicalized_lanelet_poses.end(),
            std::back_inserter(canonicalized_all));
        }
      }
      return canonicalized_all;
    } else {
      return {};
    }
  };

  /// @note If s value unders 0, it means this pose is on the previous lanelet.
  if (lanelet_pose.s < 0) {
    return canonicalize_to_previous_lanelet(lanelet_pose);
  }
  /// @note If s value overs it's lanelet length, it means this pose is on the next lanelet.
  else if (lanelet_pose.s > (getLaneletLength(lanelet_pose.lanelet_id))) {
    return canonicalize_to_next_lanelet(lanelet_pose);
  }
  /// @note If s value is in range [0,length_of_the_lanelet], return empty.
  else {
    /// Review comment, please remove before next code review. => Is this return value is valid? I think {lanelet_pose}; is a valid return value.
    return {};
  }
```

hakuturu583 replied:
I left review comment `Review comment, please remove before next code review. => Is this return value is valid? I think {lanelet_pose}; is a valid return value.`. Please check.

lchojnack replied:
I've applied the suggested change -> d88d8b60c53d2c3e5cf66cb691f802ad4812b507
I've also added additional testcases for s in range [s, length_of_lanelet]

hakuturu583 commented on simulation/traffic_simulator/src/data_type/lanelet_pose.cpp:
```diff
@@ -76,6 +80,26 @@ auto CanonicalizedLaneletPose::canonicalize(
+  std::vector<std::int64_t> shortest_route =
+    hdmap_utils->getRoute(from.lanelet_id, lanelet_poses_[0].lanelet_id);
+  LaneletPose alternative_lanelet_pose = lanelet_poses_[0];
+  for (const auto laneletPose : lanelet_poses_) {
+    auto route = hdmap_utils->getRoute(from.lanelet_id, laneletPose.lanelet_id);
+    if (shortest_route.size() > route.size()) {
+      shortest_route = route;
+      alternative_lanelet_pose = laneletPose;
+    }
```
```suggestion
  for (const auto & lanelet_pose : lanelet_poses_) {
    const auto route = hdmap_utils->getRoute(from.lanelet_id, lanelet_pose.lanelet_id);
    if (shortest_route.size() > route.size()) {
      shortest_route = route;
      alternative_lanelet_pose = lanelet_pose;
    }
  }
```

hakuturu583 replied:
Please add const and & if possible.

lchojnack replied:
I've applied the suggested change -> f0ef51bb33af1555161b5472b57285bdfec7a6ca

hakuturu583 commented on simulation/behavior_tree_plugin/src/pedestrian/pedestrian_action_node.cpp:
```diff
@@ -91,7 +91,40 @@ auto PedestrianActionNode::estimateLaneletPose(const geometry_msgs::msg::Pose &
+    if (
+      const auto canonicalized_lanelet_pose =
+        std::get<std::optional<traffic_simulator::LaneletPose>>(canonicalized)) {
+      // If canonicalize succeed, set canonicalized pose and set other values.
```
```suggestion
      /// @note If canonicalize succeed, set canonicalized pose and set other values.
```

lchojnack replied:
I've applied suggested code change -> 3dbef4062aacf73950d8d32f41082e451b4f8f34

hakuturu583 commented on simulation/behavior_tree_plugin/src/pedestrian/pedestrian_action_node.cpp:
```diff
@@ -91,7 +91,40 @@ auto PedestrianActionNode::estimateLaneletPose(const geometry_msgs::msg::Pose &
+      // If canonicalize succeed, set canonicalized pose and set other values.
+      return traffic_simulator::CanonicalizedLaneletPose(lanelet_pose.value(), hdmap_utils);
+    } else {
+      // If canonicalize failed, set end of road lanelet pose.
```
```suggestion
      /// @note If canonicalize failed, set end of road lanelet pose.
```

lchojnack replied:
I've applied suggested code change -> 3dbef4062aacf73950d8d32f41082e451b4f8f34

hakuturu583 commented on simulation/behavior_tree_plugin/src/pedestrian/pedestrian_action_node.cpp:
```diff
@@ -91,7 +91,40 @@ auto PedestrianActionNode::estimateLaneletPose(const geometry_msgs::msg::Pose &
+      if (
+        const auto end_of_road_lanelet_id = std::get<std::optional<std::int64_t>>(canonicalized)) {
+        if (lanelet_pose.value().s < 0) {
+          traffic_simulator::LaneletPose end_of_road_lanelet_pose;
+          {
+            end_of_road_lanelet_pose.lanelet_id = end_of_road_lanelet_id.value();
+            end_of_road_lanelet_pose.s = 0;
+            end_of_road_lanelet_pose.offset = lanelet_pose.value().offset;
+            end_of_road_lanelet_pose.rpy = lanelet_pose.value().rpy;
+          }
+          return traffic_simulator::CanonicalizedLaneletPose(end_of_road_lanelet_pose, hdmap_utils);
```
Please use build<messagetype> function for building ros 2 message.

```suggestion
          return traffic_simulator::CanonicalizedLaneletPose(
            traffic_simulator_msgs::build<traffic_simulator::LaneletPose>()
              .lanelet_id(end_of_road_lanelet_id.value())
              .s(0.0)
              .offset(lanelet_pose.value().offset)
              .rpy(lanelet_pose.value().rpy),
            hdmap_utils);
```

lchojnack replied:
I've applied the suggested code change -> 890745050daf79d355861104b717c035dacaabf9

hakuturu583 commented on simulation/behavior_tree_plugin/src/pedestrian/pedestrian_action_node.cpp:
```diff
@@ -91,7 +91,40 @@ auto PedestrianActionNode::estimateLaneletPose(const geometry_msgs::msg::Pose &
+          }
+          return traffic_simulator::CanonicalizedLaneletPose(end_of_road_lanelet_pose, hdmap_utils);
+        } else {
+          traffic_simulator::LaneletPose end_of_road_lanelet_pose;
+          {
+            end_of_road_lanelet_pose.lanelet_id = end_of_road_lanelet_id.value();
+            end_of_road_lanelet_pose.s =
+              hdmap_utils->getLaneletLength(end_of_road_lanelet_id.value());
+            end_of_road_lanelet_pose.offset = lanelet_pose.value().offset;
+            end_of_road_lanelet_pose.rpy = lanelet_pose.value().rpy;
+          }
+          return traffic_simulator::CanonicalizedLaneletPose(end_of_road_lanelet_pose, hdmap_utils);
```
Please use build function for building ros 2 message.
```suggestion
          return traffic_simulator::CanonicalizedLaneletPose(
            traffic_simulator_msgs::build<traffic_simulator::LaneletPose>()
              .lanelet_id(end_of_road_lanelet_id.value())
              .s(hdmap_utils->getLaneletLength(end_of_road_lanelet_id.value()))
              .offset(lanelet_pose.value().offset)
              .rpy(lanelet_pose.value().rpy),
            hdmap_utils);
```

lchojnack replied:
I've applied the suggested code change -> 890745050daf79d355861104b717c035dacaabf9

lchojnack commented:
@hakuturu583 I've applied the lastest review feedback. The PR is ready for re-review.
